File: WebDriverTests_Java/WebDriverTests/src/test/java/core/systemPages/WebDriverFunctionalTestBase.java
package core.systemPages;
import core.webDriver.WebDriverCache;
import core.systemControls.PageLoadCounter;
import org.testng.annotations.AfterClass;
import org.testng.annotations.BeforeClass;
import java.net.MalformedURLException;
import java.net.URL;

public abstract class WebDriverFunctionalTestBase
{
    @BeforeClass
    public void SetUp()
    {
        WebDriverCache.RestartIfNeed();
        WebDriverCache.getWebDriver().DeleteAllCookies();
        WebDriverCache.getWebDriver().CleanDownloadDirectory();
        PageLoadCounter.Reset();
    }

    @AfterClass
    public void TearDown()
    {
        try
        {
            if (WebDriverCache.getInitialized())
            {
                TearDownInternal();
            }
        }
        finally
        {
            WebDriverCache.getWebDriver().DeleteAllCookies();
            WebDriverCache.getWebDriver().ExecuteScript("localStorage.clear();");
            WebDriverCache.DestroyInstance();
        }
    }

    private void TearDownInternal()
    {
        try
        {
            PageLoadCounter.AssertPageNotLoaded();
        }
        finally
        {
        }
    }

    public final <TPage extends PageBase> TPage LoadPage(Class<TPage> pageClass, String localPath)  {
        URL g = null;
        try {
            g = new URL(new URL(String.format("http://%1$s/", getApplicationBaseUrl())), localPath);
        } catch (MalformedURLException e) {
            e.printStackTrace();
        }
        return PageBase.GoToUri(pageClass,g);
    }

    public final <TPage extends PageBase> TPage LoadPageUrl(Class<TPage> pageClass, String localPath)  {
        try {
            return PageBase.<TPage>GoToUri(pageClass,new URL(localPath));
        } catch (MalformedURLException e) {
            e.printStackTrace();
        }
        return null;
    }

    public String getApplicationBaseUrl()
    {
        return "stage.ddelivery.ru";
    }

    private static void CaptureJavascriptErrors()
    {
        Object tempVar = WebDriverCache.getWebDriver().ExecuteScript("return window.jsErrors");
        String errors = (String)((tempVar instanceof String) ? tempVar : null);
    }
}

File: WebDriverTests_Csharp/Autotest.Core/Autotests.Utilities/WebTestCore/WebDriverFunctionalTestBase.cs
using System;
using Autotests.Utilities.Core.Web.Controllers;
using Autotests.Utilities.WebTestCore.Pages;
using Autotests.Utilities.WebTestCore.SystemControls;
using Autotests.Utilities.WebTestCore.TestSystem;
using Autotests.Utilities.WebTestCore.Utils;
using NUnit.Framework;

namespace Autotests.Utilities.WebTestCore
{
    [TestFixture]
    public abstract class WebDriverFunctionalTestBase
    {
        [SetUp]
        public virtual void SetUp()
        {
            Log4NetConfiguration.InitializeOnce();
            WebDriverCache.RestartIfNeed();
            WebDriverCache.WebDriver.DeleteAllCookies();
            WebDriverCache.WebDriver.CleanDownloadDirectory();
            PageLoadCounter.Reset();
            JsLogger.Reset();
        }

        [TearDown]
        public virtual void TearDown()
        {
            try
            {
                if (WebDriverCache.Initialized)
                    TearDownInternal();
            }
            finally
            {
                FieldsCleanerCache.Clean(this);
            }
        }

        private void TearDownInternal()
        {
            try
            {
                PageLoadCounter.AssertPageNotLoaded();
            }
            finally
            {
                TestContext testContext = TestContext.CurrentContext;
                try
                {
                    if (testContext.Result.Status == TestStatus.Failed)
                    {
                        CaptureJavascriptErrors();
                        JsLogger.Show();
//                        WebDriverCache.WebDriver.CaptureScreenshot();
                    }
                }
                catch
                {
                    Console.Out.WriteLine("Cannot save screenshot. Probably TearDown method called not in NUnit context");
                }
            }
        }

        protected TPage LoadPage<TPage>(string localPath) where TPage : PageBase, new()
        {
            return PageBase.GoToUri<TPage>(new Uri(new Uri(string.Format("http://{0}/", ApplicationBaseUrl)), localPath));
        }

        protected TPage LoadPageUrl<TPage>(string localPath) where TPage : PageBase, new()
        {
            return PageBase.GoToUri<TPage>(new Uri(localPath));
        }

        public virtual string ApplicationBaseUrl { get { return "stage.ddelivery.ru"; } }

        private static void CaptureJavascriptErrors()
        {
            var errors = WebDriverCache.WebDriver.ExecuteScript("return window.jsErrors") as string;
            if (!string.IsNullOrEmpty(errors))
                Console.WriteLine("Javascript errors:\n" + errors);
        }

    }
}
